File: protocols/proven-coap/ffi/zig/src/coap.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// coap.zig -- Zig FFI implementation of proven-coap.
//
// Implements the CoAP (RFC 7252) server endpoint state machine with:
//   - 64-slot mutex-protected endpoint pool
//   - Resource registration per endpoint (max 32 resources)
//   - Observation tracking per resource (RFC 7641, max 64 observers)
//   - Method bitmask per resource (GET=0x01, POST=0x02, PUT=0x04, DELETE=0x08)
//   - Request handling with method validation
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching CoAPABI.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching CoAPABI.Types.idr tag assignments)
// =========================================================================

/// CoAP request methods (ABI tags 0-3).
pub const Method = enum(u8) {
    get = 0,
    post = 1,
    put = 2,
    delete = 3,
};

/// CoAP message types (ABI tags 0-3).
pub const MessageType = enum(u8) {
    confirmable = 0,
    non_confirmable = 1,
    acknowledgement = 2,
    reset = 3,
};

/// CoAP content formats (ABI tags 0-6).
pub const ContentFormat = enum(u8) {
    text_plain = 0,
    link_format = 1,
    xml = 2,
    octet_stream = 3,
    exi = 4,
    json = 5,
    cbor = 6,
};

/// CoAP response class (ABI tags 0-4).
pub const ResponseClass = enum(u8) {
    success = 0,
    client_error = 1,
    server_error = 2,
    signaling = 3,
    empty = 4,
};

/// CoAP server endpoint lifecycle states (ABI tags 0-4).
pub const SessionState = enum(u8) {
    idle = 0,
    bound = 1,
    serving = 2,
    observing = 3,
    shutdown = 4,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent endpoints.
const MAX_ENDPOINTS: usize = 64;

/// Maximum resources per endpoint.
const MAX_RESOURCES: usize = 32;

/// Maximum observers per endpoint (across all resources).
const MAX_OBSERVERS: usize = 64;

/// Maximum URI path length in bytes.
const MAX_PATH_LEN: usize = 256;

/// A registered CoAP resource.
const Resource = struct {
    /// URI path (e.g., "/temperature", "/light").
    path: [MAX_PATH_LEN]u8,
    path_len: u32,
    /// Allowed methods bitmask: bit0=GET, bit1=POST, bit2=PUT, bit3=DELETE.
    methods: u8,
    /// Whether this resource slot is active.
    active: bool,
};

/// An observation subscription (RFC 7641).
const Observer = struct {
    /// URI path of the observed resource.
    path: [MAX_PATH_LEN]u8,
    path_len: u32,
    /// Client token for matching responses.
    token: u64,
    /// Whether this observer slot is active.
    active: bool,
};

/// A CoAP server endpoint.
const Endpoint = struct {
    /// Current endpoint lifecycle state.
    state: SessionState,
    /// Bound UDP port.
    port: u16,
    /// Registered resources.
    resources: [MAX_RESOURCES]Resource,
    /// Number of active resources.
    resource_count: u32,
    /// Observation subscriptions.
    observers: [MAX_OBSERVERS]Observer,
    /// Number of active observers.
    observer_count: u32,
    /// Total requests handled (monotonic counter).
    requests_handled: u64,
    /// Whether this endpoint slot is in use.
    active: bool,
};

/// Default (empty) resource.
const empty_resource: Resource = .{
    .path = [_]u8{0} ** MAX_PATH_LEN,
    .path_len = 0,
    .methods = 0,
    .active = false,
};

/// Default (empty) observer.
const empty_observer: Observer = .{
    .path = [_]u8{0} ** MAX_PATH_LEN,
    .path_len = 0,
    .token = 0,
    .active = false,
};

/// Default (empty) endpoint.
const empty_endpoint: Endpoint = .{
    .state = .idle,
    .port = 0,
    .resources = [_]Resource{empty_resource} ** MAX_RESOURCES,
    .resource_count = 0,
    .observers = [_]Observer{empty_observer} ** MAX_OBSERVERS,
    .observer_count = 0,
    .requests_handled = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var endpoints: [MAX_ENDPOINTS]Endpoint = [_]Endpoint{empty_endpoint} ** MAX_ENDPOINTS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_ENDPOINTS) return null;
    const idx: usize = @intCast(slot);
    if (!endpoints[idx].active) return null;
    return idx;
}

/// Find a resource by path within an endpoint.
fn findResource(idx: usize, path: []const u8) ?usize {
    for (&endpoints[idx].resources, 0..) |*r, i| {
        if (r.active and r.path_len == path.len and
            std.mem.eql(u8, r.path[0..r.path_len], path))
        {
            return i;
        }
    }
    return null;
}

/// Check if a method tag is valid (0-3).
fn validMethod(method: u8) bool {
    return method <= 3;
}

/// Convert a method tag to its bitmask position.
fn methodBit(method: u8) u8 {
    return @as(u8, 1) << @intCast(method);
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

// -- ABI version --------------------------------------------------------------

/// Returns the ABI version number. Must match Foreign.abiVersion in Idris2.
pub export fn coap_abi_version() callconv(.c) u32 {
    return 1;
}

// -- Lifecycle ----------------------------------------------------------------

/// Create a new CoAP endpoint. Returns slot index (>=0) or -1 on failure.
/// The endpoint starts in Bound state (Idle -> Bound transition applied).
pub export fn coap_create(port: u16) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (port == 0) return -1;

    for (&endpoints, 0..) |*ep, i| {
        if (!ep.active) {
            ep.* = empty_endpoint;
            ep.port = port;
            ep.state = .bound; // Idle -> Bound
            ep.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy an endpoint, releasing its slot.
pub export fn coap_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_ENDPOINTS) return;
    endpoints[@intCast(slot)] = empty_endpoint;
}

// -- State queries ------------------------------------------------------------

/// Returns the current SessionState tag for an endpoint.
pub export fn coap_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0; // idle fallback
    return @intFromEnum(endpoints[idx].state);
}

/// Returns 1 if the endpoint can serve requests, 0 otherwise.
pub export fn coap_can_serve(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    const state = endpoints[idx].state;
    return if (state == .serving or state == .observing) 1 else 0;
}

// -- Resource management ------------------------------------------------------

/// Register a resource at a URI path. Returns 0 on success, 1 on rejection.
/// Transitions: Bound -> Serving, or stays Serving/Observing.
pub export fn coap_register_resource(
    slot: c_int,
    path_ptr: [*]const u8,
    path_len: u32,
    methods: u8,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = endpoints[idx].state;
    if (state != .bound and state != .serving and state != .observing) return 1;
    if (path_len == 0 or path_len > MAX_PATH_LEN) return 1;
    if (methods == 0 or methods > 0x0F) return 1; // must have at least one method, max 4 bits

    const path = path_ptr[0..path_len];

    // Check for duplicate resource path
    if (findResource(idx, path) != null) return 1;

    // Find a free resource slot
    for (&endpoints[idx].resources) |*r| {
        if (!r.active) {
            @memcpy(r.path[0..path_len], path);
            r.path_len = path_len;
            r.methods = methods;
            r.active = true;
            endpoints[idx].resource_count += 1;
            if (endpoints[idx].state == .bound) {
                endpoints[idx].state = .serving;
            }
            return 0;
        }
    }
    return 1;
}

/// Unregister a resource by path. Returns 0 on success, 1 on rejection.
/// May transition Serving -> Bound if last resource.
pub export fn coap_unregister_resource(
    slot: c_int,
    path_ptr: [*]const u8,
    path_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (path_len == 0 or path_len > MAX_PATH_LEN) return 1;

    const path = path_ptr[0..path_len];
    const ri = findResource(idx, path) orelse return 1;

    // Cancel all observers for this resource
    for (&endpoints[idx].observers) |*obs| {
        if (obs.active and obs.path_len == path_len and
            std.mem.eql(u8, obs.path[0..obs.path_len], path))
        {
            obs.active = false;
            obs.path_len = 0;
            endpoints[idx].observer_count -= 1;
        }
    }

    endpoints[idx].resources[ri].active = false;
    endpoints[idx].resource_count -= 1;

    // If no resources remain, transition to Bound
    if (endpoints[idx].resource_count == 0) {
        if (endpoints[idx].state == .serving or endpoints[idx].state == .observing) {
            endpoints[idx].state = .bound;
        }
    } else if (endpoints[idx].observer_count == 0 and endpoints[idx].state == .observing) {
        endpoints[idx].state = .serving;
    }

    return 0;
}

/// Returns the number of registered resources for an endpoint.
pub export fn coap_resource_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return endpoints[idx].resource_count;
}

// -- Request handling ---------------------------------------------------------

/// Handle an incoming request. Returns ResponseClass tag.
/// Validates method against resource's allowed methods bitmask.
pub export fn coap_handle_request(
    slot: c_int,
    method: u8,
    msg_type: u8,
    msg_id: u16,
    token_len: u8,
    path_ptr: [*]const u8,
    path_len: u32,
    payload_ptr: [*]const u8,
    payload_len: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    _ = msg_type;
    _ = msg_id;
    _ = token_len;
    _ = payload_ptr;
    _ = payload_len;

    const idx = validSlot(slot) orelse return @intFromEnum(ResponseClass.server_error);
    const state = endpoints[idx].state;
    if (state != .serving and state != .observing) {
        return @intFromEnum(ResponseClass.server_error);
    }
    if (!validMethod(method)) return @intFromEnum(ResponseClass.client_error);
    if (path_len == 0 or path_len > MAX_PATH_LEN) return @intFromEnum(ResponseClass.client_error);

    const path = path_ptr[0..path_len];
    const ri = findResource(idx, path) orelse return @intFromEnum(ResponseClass.client_error); // 4.04 Not Found

    // Check if the method is allowed for this resource
    const bit = methodBit(method);
    if (endpoints[idx].resources[ri].methods & bit == 0) {
        return @intFromEnum(ResponseClass.client_error); // 4.05 Method Not Allowed
    }

    endpoints[idx].requests_handled += 1;
    return @intFromEnum(ResponseClass.success); // 2.xx
}

// -- Observation (RFC 7641) ---------------------------------------------------

/// Add an observer for a resource. Returns 0 on success, 1 on rejection.
/// Transitions: Serving -> Observing, or stays Observing.
pub export fn coap_add_observer(
    slot: c_int,
    path_ptr: [*]const u8,
    path_len: u32,
    token: u64,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = endpoints[idx].state;
    if (state != .serving and state != .observing) return 1;
    if (path_len == 0 or path_len > MAX_PATH_LEN) return 1;

    const path = path_ptr[0..path_len];

    // Verify resource exists
    if (findResource(idx, path) == null) return 1;

    // Find a free observer slot
    for (&endpoints[idx].observers) |*obs| {
        if (!obs.active) {
            @memcpy(obs.path[0..path_len], path);
            obs.path_len = path_len;
            obs.token = token;
            obs.active = true;
            endpoints[idx].observer_count += 1;
            endpoints[idx].state = .observing;
            return 0;
        }
    }
    return 1;
}

/// Remove an observer. Returns 0 on success, 1 on rejection.
/// May transition Observing -> Serving if last observer.
pub export fn coap_remove_observer(
    slot: c_int,
    path_ptr: [*]const u8,
    path_len: u32,
    token: u64,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (path_len == 0 or path_len > MAX_PATH_LEN) return 1;

    const path = path_ptr[0..path_len];

    for (&endpoints[idx].observers) |*obs| {
        if (obs.active and obs.token == token and obs.path_len == path_len and
            std.mem.eql(u8, obs.path[0..obs.path_len], path))
        {
            obs.active = false;
            obs.path_len = 0;
            endpoints[idx].observer_count -= 1;

            if (endpoints[idx].observer_count == 0 and endpoints[idx].state == .observing) {
                endpoints[idx].state = .serving;
            }
            return 0;
        }
    }
    return 1;
}

/// Returns the total number of active observers.
pub export fn coap_observer_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return endpoints[idx].observer_count;
}

/// Notify all observers of a resource. Returns number of observers notified.
pub export fn coap_notify_observers(
    slot: c_int,
    path_ptr: [*]const u8,
    path_len: u32,
) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 0;
    if (path_len == 0 or path_len > MAX_PATH_LEN) return 0;

    const path = path_ptr[0..path_len];
    var notified: u32 = 0;

    for (&endpoints[idx].observers) |*obs| {
        if (obs.active and obs.path_len == path_len and
            std.mem.eql(u8, obs.path[0..obs.path_len], path))
        {
            notified += 1;
        }
    }
    return notified;
}

// -- Shutdown / Cleanup -------------------------------------------------------

/// Shutdown the endpoint. Returns 0 on success, 1 on rejection.
pub export fn coap_shutdown(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = endpoints[idx].state;
    if (state == .bound or state == .serving or state == .observing) {
        endpoints[idx].state = .shutdown;
        return 0;
    }
    return 1;
}

/// Complete cleanup after shutdown. Returns 0 on success, 1 on rejection.
pub export fn coap_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (endpoints[idx].state != .shutdown) return 1;

    endpoints[idx].state = .idle;
    endpoints[idx].resources = [_]Resource{empty_resource} ** MAX_RESOURCES;
    endpoints[idx].resource_count = 0;
    endpoints[idx].observers = [_]Observer{empty_observer} ** MAX_OBSERVERS;
    endpoints[idx].observer_count = 0;

    return 0;
}

// -- Stateless transition table -----------------------------------------------

/// Check if an endpoint state transition is valid.
pub export fn coap_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Idle -> Bound
    if (from == 1 and to == 2) return 1; // Bound -> Serving
    if (from == 2 and to == 2) return 1; // Serving -> Serving (add more resources)
    if (from == 2 and to == 1) return 1; // Serving -> Bound (all resources removed)
    if (from == 2 and to == 3) return 1; // Serving -> Observing
    if (from == 3 and to == 3) return 1; // Observing -> Observing
    if (from == 3 and to == 2) return 1; // Observing -> Serving (all observers removed)
    if (from == 1 and to == 4) return 1; // Bound -> Shutdown
    if (from == 2 and to == 4) return 1; // Serving -> Shutdown
    if (from == 3 and to == 4) return 1; // Observing -> Shutdown
    if (from == 4 and to == 0) return 1; // Shutdown -> Idle
    return 0;
}

File: bindings/elixir/lib/proven_servers/coap.ex
# SPDX-License-Identifier: MPL-2.0
# Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
defmodule ProvenServers.Coap do
  @moduledoc """
  CoAP (Constrained Application Protocol) types for the proven-servers ABI.
  
  Mirrors the Idris2 module `CoapABI.Types` and its type definitions:
  - `Method`        — CoAP request methods (4 constructors, tags 0-3)
  - `MessageType`   — CoAP message types (4 constructors, tags 0-3)
  - `ContentFormat` — CoAP content formats (7 constructors, tags 0-6)
  - `ResponseClass` — CoAP response class codes (5 constructors, tags 0-4)
  - `SessionState`  — CoAP server lifecycle (5 constructors, tags 0-4)
  
  All discriminant values match the Idris2 ABI tag definitions exactly.

  All tag values match the Idris2 ABI definitions exactly.
  """

  @doc "Standard CoAP port (RFC 7252)."
  @spec coap_port() :: non_neg_integer()
  def coap_port, do: 5683

  @doc "Standard CoAPS (CoAP over DTLS) port (RFC 7252)."
  @spec coaps_port() :: non_neg_integer()
  def coaps_port, do: 5684

  @doc "Default CoAP block size (RFC 7959)."
  @spec coap_default_block_size() :: non_neg_integer()
  def coap_default_block_size, do: 1024

  # ===========================================================================
  # Method (tags 0-3)
  # ===========================================================================

  @typedoc """
  Method types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type method :: :get | :post | :put | :delete

  @method_tags %{
    get: 0,
    post: 1,
    put: 2,
    delete: 3,
  }

  @tag_to_method Map.new(@method_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `Method` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Coap.method_from_tag(0)
      {:ok, :get}
  """
  @spec method_from_tag(non_neg_integer()) :: {:ok, method()} | :error
  def method_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_method, tag)}
  end

  def method_from_tag(_tag), do: :error

  @doc """
  Encode a `Method` to the C-ABI tag value.
  """
  @spec method_to_tag(method()) :: non_neg_integer()
  def method_to_tag(val) when is_map_key(@method_tags, val) do
    Map.fetch!(@method_tags, val)
  end

  @doc """
  All `Method` variants in tag order.
  """
  @spec all_methods() :: [method()]
  def all_methods, do: [:get, :post, :put, :delete]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this method is safe (does not alter server state).
  """
  @spec is_safe?(method()) :: boolean()
  def is_safe?(val) when val in [:get], do: true
  def is_safe?(_val), do: false

  @doc """
  Whether this method is idempotent.
  """
  @spec is_idempotent?(method()) :: boolean()
  def is_idempotent?(val) when val in [:get, :put, :delete], do: true
  def is_idempotent?(_val), do: false

  # ===========================================================================
  # MessageType (tags 0-3)
  # ===========================================================================

  @typedoc """
  MessageType types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type message_type :: :confirmable | :non_confirmable | :acknowledgement | :reset

  @message_type_tags %{
    confirmable: 0,
    non_confirmable: 1,
    acknowledgement: 2,
    reset: 3,
  }

  @tag_to_message_type Map.new(@message_type_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `MessageType` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..3, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Coap.message_type_from_tag(0)
      {:ok, :confirmable}
  """
  @spec message_type_from_tag(non_neg_integer()) :: {:ok, message_type()} | :error
  def message_type_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 3 do
    {:ok, Map.fetch!(@tag_to_message_type, tag)}
  end

  def message_type_from_tag(_tag), do: :error

  @doc """
  Encode a `MessageType` to the C-ABI tag value.
  """
  @spec message_type_to_tag(message_type()) :: non_neg_integer()
  def message_type_to_tag(val) when is_map_key(@message_type_tags, val) do
    Map.fetch!(@message_type_tags, val)
  end

  @doc """
  All `MessageType` variants in tag order.
  """
  @spec all_message_types() :: [message_type()]
  def all_message_types, do: [:confirmable, :non_confirmable, :acknowledgement, :reset]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this message type requires a response.
  """
  @spec requires_response?(message_type()) :: boolean()
  def requires_response?(val) when val in [:confirmable], do: true
  def requires_response?(_val), do: false

  @doc """
  Whether this message type is a response.
  """
  @spec is_response?(message_type()) :: boolean()
  def is_response?(val) when val in [:acknowledgement, :reset], do: true
  def is_response?(_val), do: false

  # ===========================================================================
  # ContentFormat (tags 0-6)
  # ===========================================================================

  @typedoc """
  ContentFormat types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type content_format ::
          :text_plain
          | :link_format
          | :xml
          | :octet_stream
          | :exi
          | :json
          | :cbor

  @content_format_tags %{
    text_plain: 0,
    link_format: 1,
    xml: 2,
    octet_stream: 3,
    exi: 4,
    json: 5,
    cbor: 6,
  }

  @tag_to_content_format Map.new(@content_format_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ContentFormat` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..6, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Coap.content_format_from_tag(0)
      {:ok, :text_plain}
  """
  @spec content_format_from_tag(non_neg_integer()) :: {:ok, content_format()} | :error
  def content_format_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 6 do
    {:ok, Map.fetch!(@tag_to_content_format, tag)}
  end

  def content_format_from_tag(_tag), do: :error

  @doc """
  Encode a `ContentFormat` to the C-ABI tag value.
  """
  @spec content_format_to_tag(content_format()) :: non_neg_integer()
  def content_format_to_tag(val) when is_map_key(@content_format_tags, val) do
    Map.fetch!(@content_format_tags, val)
  end

  @doc """
  All `ContentFormat` variants in tag order.
  """
  @spec all_content_formats() :: [content_format()]
  def all_content_formats, do: [:text_plain, :link_format, :xml, :octet_stream, :exi, :json, :cbor]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  The IANA media type string for this content format.
        match self {

  Whether this format is text-based (human-readable).
  """
  @spec is_text_based?(content_format()) :: boolean()
  def is_text_based?(val) when val in [:text_plain, :link_format, :xml, :json], do: true
  def is_text_based?(_val), do: false

  # ===========================================================================
  # ResponseClass (tags 0-4)
  # ===========================================================================

  @typedoc """
  ResponseClass types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type response_class :: :success | :client_error | :server_error | :signaling | :empty

  @response_class_tags %{
    success: 0,
    client_error: 1,
    server_error: 2,
    signaling: 3,
    empty: 4,
  }

  @tag_to_response_class Map.new(@response_class_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `ResponseClass` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Coap.response_class_from_tag(0)
      {:ok, :success}
  """
  @spec response_class_from_tag(non_neg_integer()) :: {:ok, response_class()} | :error
  def response_class_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_response_class, tag)}
  end

  def response_class_from_tag(_tag), do: :error

  @doc """
  Encode a `ResponseClass` to the C-ABI tag value.
  """
  @spec response_class_to_tag(response_class()) :: non_neg_integer()
  def response_class_to_tag(val) when is_map_key(@response_class_tags, val) do
    Map.fetch!(@response_class_tags, val)
  end

  @doc """
  All `ResponseClass` variants in tag order.
  """
  @spec all_response_classs() :: [response_class()]
  def all_response_classs, do: [:success, :client_error, :server_error, :signaling, :empty]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether this response class indicates success.
  """
  @spec is_success?(response_class()) :: boolean()
  def is_success?(val) when val in [:success], do: true
  def is_success?(_val), do: false

  @doc """
  Whether this response class indicates an error.
  """
  @spec is_error?(response_class()) :: boolean()
  def is_error?(val) when val in [:client_error, :server_error], do: true
  def is_error?(_val), do: false

  # ===========================================================================
  # SessionState (tags 0-4)
  # ===========================================================================

  @typedoc """
  SessionState types.

  Tag values match the Idris2 ABI definitions exactly.
  """
  @type session_state :: :idle | :bound | :serving | :observing | :shutdown

  @session_state_tags %{
    idle: 0,
    bound: 1,
    serving: 2,
    observing: 3,
    shutdown: 4,
  }

  @tag_to_session_state Map.new(@session_state_tags, fn {k, v} -> {v, k} end)

  @doc """
  Decode a `SessionState` from the C-ABI tag value.

  Returns `{:ok, atom}` for valid tags 0..4, `:error` for invalid.

  ## Examples

      iex> ProvenServers.Coap.session_state_from_tag(0)
      {:ok, :idle}
  """
  @spec session_state_from_tag(non_neg_integer()) :: {:ok, session_state()} | :error
  def session_state_from_tag(tag) when is_integer(tag) and tag >= 0 and tag <= 4 do
    {:ok, Map.fetch!(@tag_to_session_state, tag)}
  end

  def session_state_from_tag(_tag), do: :error

  @doc """
  Encode a `SessionState` to the C-ABI tag value.
  """
  @spec session_state_to_tag(session_state()) :: non_neg_integer()
  def session_state_to_tag(val) when is_map_key(@session_state_tags, val) do
    Map.fetch!(@session_state_tags, val)
  end

  @doc """
  All `SessionState` variants in tag order.
  """
  @spec all_session_states() :: [session_state()]
  def all_session_states, do: [:idle, :bound, :serving, :observing, :shutdown]

  @doc """
  Decode from an ABI tag value.

  Encode to the ABI tag value.

  Whether the server is ready to handle requests.
  """
  @spec is_active?(session_state()) :: boolean()
  def is_active?(val) when val in [:serving, :observing], do: true
  def is_active?(_val), do: false

end
